# Cross-Fitting and $p$-Value Combination

In this tutorial, we will demonstrate how to implement cross-fitting and $p$-value combination by using `Crosser`. The `Crosser` class provides a user-friendly interface to conduct nested cross-validation (CV) that results in a more stable result under small sample sizes.

To install `tofi` via `pip`, please use:
```
pip install pytofi
```

We will again use the `diabetes` data set. The following code prepares the data for the tutorial.

In [1]:
from sklearn import datasets
from sklearn.model_selection import train_test_split
X, y = datasets.load_diabetes(
    return_X_y = True,
    as_frame = True,
    scaled = False)

## `Crosser` Overview

Briefly speaking, `Crosser` is an object that can conduct nested cross-validation (CV) to improve the stability of learning result. Suppose we have split the a data set $Z$ into 5 folds, say $Z^{(1)}, Z^{(2)}, Z^{(3)}, Z^{(4)}, Z^{(5)}$. For each $k \in \{1,2,3,4,5\}$, we set $Z_\text{train}^{(k)} = \{Z_l\}_{l \neq k}$ and $Z_\text{test}^{(k)} = \{Z_k\}$. Then $Z_\text{train}^{(k)}$ is used to train $\widehat{f}^{(k)}$ and $Z_\text{test}^{(k)}$ is used to evaluate the performance of $\widehat{f}^{(k)}$. Note that when training $\widehat{f}^{(k)}$ an inner CV loop may be used.

To understand `Crosser` more concretely, let us create a cross-validator via `KFold`. The cv can help us to create the training and test indices for 5-fold CV. An important thing here is that a specific `random_state` must be given. Otherwise, `tofi` may yield a wrong result when making statistical inferences.

In [2]:
from sklearn.model_selection import KFold
cv = KFold(
    n_splits = 5,
    shuffle = True,
    random_state = 1)

Now we initialize a `Crosser` object. When initializing, an estimator and a cross-validator must be given. The estimator can be a usual `scikit-learn` estimator or a `GridSearchCV` object as in this example. The `GridSearchCV` object runs an inner CV loop to find an optimal tuning parameter value for each $\widehat{f}^{(k)}$. The `fit` method trains $\widehat{f}^{(1)}$, $\widehat{f}^{(2)}$, ..,$\widehat{f}^{(5)}$ on their corresponding training sets.

In [3]:
from tofi import Crosser
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestRegressor
learner = Crosser(
    GridSearchCV(
        estimator = RandomForestRegressor(), 
        param_grid = {
            "max_features": [3, 6, 9]}),
    cv = cv)
_ = learner.fit(X, y)

To extract the trained estimators, we can use the `estimator_` attribute:

In [4]:
len(learner.estimators_)

5

The overall performances of these trained estimators can be shown by the `summarize` method. Three types of error scores are reported. The `val_score` is the validation error extracted from the `GridSearchCV`. The `train_score` and `test_score` are calculated by `Crosser`. By default, regression tasks use `r2` scorer and classification tasks uses `accuracy`. The scoring method can be explicitly specified via the `scoring` argument when initializing a `Crosser`. Available scoring methods can be found [here](https://scikit-learn.org/stable/modules/model_evaluation.html#common-cases-predefined-values). 

In [5]:
learner.summarize()

Crosser Summary (cross_fit=True, combine=False)
 + Estimator: GridSearchCV
 + Cross-Validator: KFold (n_folds=5, n_repeats=1)
 + Train/Test Scorer: R2 (reverse=False)


val_score  train_score  test_score
repeat                                    
0        0.437874     0.922354    0.424619

By default, the `summarize` method averages the error scores across all folds. It is also possible to show the error scores for each fold by setting `cross_fit = False`. We can see that the test errors are quite different across the folds.

In [6]:
learner.summarize(
    cross_fit = False)

Crosser Summary (cross_fit=False, combine=False)
 + Estimator: GridSearchCV
 + Cross-Validator: KFold (n_folds=5, n_repeats=1)
 + Train/Test Scorer: R2 (reverse=False)


val_score  train_score  test_score
split repeat fold                                    
0     0      0      0.476149     0.926979    0.336819
1     0      1      0.464522     0.925863    0.324295
2     0      2      0.416937     0.920513    0.435586
3     0      3      0.412163     0.920405    0.492918
4     0      4      0.419600     0.918013    0.533475

The `Crosser` also has a `predict` method for prediction. For a regressor, by default, it makes predictions by averaging the predictions made by all the trained estimators:

In [7]:
learner.predict(
    X.iloc[:5,:])

array([178.718,  80.436, 158.13 , 189.274, 117.98 ])

When the estimator is a classifier, the prediction is based on majority voting. If we hope to make predictions based on a model trained on specific split, the argument `split` can be used:

In [8]:
learner.predict(
    X.iloc[:5,:],
    split = 0)

array([175.13,  80.43, 162.09, 189.4 ,  98.86])

## Inference with Cross-Fitting via `Crosser`

Now we will demonstrate how to implement cross-fitting strategy to improve both statistical power and stability of a test with the help of `Crosser`. In breif, the cross-fitting atrategy tries to integrate the test results made by the folds:

+ When the null distribution is constructed by `null_dist = resampling`, the integration is based on Algorithm 4 of Tansey et al (2022);
+ When the null distribution is constructed by `null_dist = normality` or `null_dist = permutation`, the integration is based on Algorithm 3 of Williamson et al. (2023).

In this example, the conditional predictive impact (CPI) is used to test the significance of `sex`. Hence, a sampler for `sex` is necessary. Note that the sampler must be trained with the same cross-validator as we specified for training the learner. Otherwise, `tofi` cannot perform cross-fitting correctly.

In [9]:
from sklearn.ensemble import RandomForestClassifier
removal = "sex"
sampler = Crosser(
    GridSearchCV(
        estimator = RandomForestClassifier(), 
        param_grid = {
            "max_features": [3, 6, 9]}),
    cv = cv)
_ = sampler.fit(
    X.drop(removal, axis = 1), 
    X[removal])

To implement cross-fitting, we replace usual estimators by their corresponding `Crosser` objects when initializing the `Inferer`. After infering, the `summarize` method will integrate the results via cross-fitting by default:

In [10]:
from tofi import CIT
cpi = CIT(
    learner, 
    sampler,
    removal,
    "CPI",
    n_copies = 100)
_ = cpi.infer(X, y)
cpi.summarize()

Inferer Summary (cross_fit=True, combine=False)
 + Method: CPI (double_split=None, perturb_size=None)
 + Null Distribution: Normality (n_copies=100, n_permutations=None)
 + Loss Function: Mean Squared Error (reverse=False)


size   estimate  std_error   p_value
removal                                      
sex       442  50.597491  19.096992  0.004031

To inspect the result on each fold, we can set `cross_fit = False`. We can see that these results are less stable and less powerful bacause of the small sample sizes of the `diabetes` data.

In [11]:
cpi.summarize(
    cross_fit = False)

Inferer Summary (cross_fit=False, combine=False)
 + Method: CPI (double_split=None, perturb_size=None)
 + Null Distribution: Normality (n_copies=100, n_permutations=None)
 + Loss Function: Mean Squared Error (reverse=False)


size    estimate  std_error   p_value
removal split repeat fold                                       
sex     0     0      0       89   -3.674510  47.449322  0.530864
        1     0      1       89  149.893611  38.727091  0.000054
        2     0      2       88   44.777228  38.528786  0.122582
        3     0      3       88   26.325489  37.592782  0.241876
        4     0      4       88   35.665637  48.097203  0.229185

## Inference with Both Cross-Fitting and $p$-Value Combination

Another way to improve the stability of testing is using $p$-value combinations. In this example, we use both cross-fitting and $p$-value combination simultaneously. We plan to conduct 5-fold cross-validation 4 times. Hence, `RepeatedKFold` is used to initialize a cross-validator:

In [12]:
from sklearn.model_selection import RepeatedKFold
cv = RepeatedKFold(
    n_splits = 5,
    n_repeats = 4,
    random_state = 1)

In `tofi`, `n_splits` is used to denote the number of all splits and the number of folds will be represented by another variable called `n_folds`. Hence, the specified `n_splits = 5` and `n_repeats = 4` in this example results in `n_splits = 20`, `n_folds = 5`, and `n_repeats = 4` that are stored in both `Crosser` and `Inferer`.

Then we train a learner and a sampler with the `RepeatedKFold` cross-validator and conduct CPI for `sex`. By default, the `summarize` method implement several p-value combination methods that aggregate the $p$-values obtained in the 4 replications.

In [13]:
learner = Crosser(
    GridSearchCV(
        estimator = RandomForestRegressor(), 
        param_grid = {
            "max_features": [3, 6, 9]}),
    cv = cv)
sampler = Crosser(
    GridSearchCV(
        estimator = RandomForestClassifier(), 
        param_grid = {
            "max_features": [3, 6, 9]}),
    cv = cv)
_ = learner.fit(X, y)
_ = sampler.fit(
    X.drop(removal, axis = 1), 
    X[removal])

In [14]:
cpi = CIT(
    learner, 
    sampler,
    removal,
    "CPI",
    n_copies = 100)
_ = cpi.infer(X, y)
cpi.summarize()

Inferer Summary (cross_fit=True, combine=True)
 + Method: CPI (double_split=None, perturb_size=None)
 + Null Distribution: Normality (n_copies=100, n_permutations=None)
 + Loss Function: Mean Squared Error (reverse=False)


size   estimate  std_error   p_value
removal method                                       
sex     gmean   442.0  48.913911  18.116957  0.009206
        median  442.0  48.913911  18.116957  0.007400
        q1      442.0  48.913911  18.116957  0.004022
        min     442.0  48.913911  18.116957  0.008031
        hmean   442.0  48.913911  18.116957  0.011197
        hommel  442.0  48.913911  18.116957  0.008380
        cauchy  442.0  48.913911  18.116957  0.002971

To see the $p$-value in each replication, we can set `combine = False` in the `summarize` method. Note that these $p$-values are still calculated based on cross-fitting. Even after cross-fitting, the $p$-values could still differ slightly across replications. That is why $p$-value combination could be helpful for enhencing stability.

In [15]:
cpi.summarize(
    combine = False)

Inferer Summary (cross_fit=True, combine=False)
 + Method: CPI (double_split=None, perturb_size=None)
 + Null Distribution: Normality (n_copies=100, n_permutations=None)
 + Loss Function: Mean Squared Error (reverse=False)


size   estimate  std_error   p_value
removal repeat                                      
sex     0        442  46.952550  16.323351  0.002011
        1        442  46.497831  18.528726  0.006045
        2        442  46.865943  18.380200  0.005389
        3        442  55.339321  19.235551  0.002008

If users are interested in the $p$-value on each split, `combine = False` and `cross_fit = False` can be used. We see that the split-level results are indeed unstable.

In [16]:
cpi.summarize(
    cross_fit = False,
    combine = False)

Inferer Summary (cross_fit=False, combine=False)
 + Method: CPI (double_split=None, perturb_size=None)
 + Null Distribution: Normality (n_copies=100, n_permutations=None)
 + Loss Function: Mean Squared Error (reverse=False)


size    estimate  std_error   p_value
removal split repeat fold                                       
sex     0     0      0       89   49.441475  35.942646  0.084478
        1     0      1       89   80.028821  26.213383  0.001133
        2     0      2       88   35.510752  38.654548  0.179134
        3     0      3       88   31.949917  34.925028  0.180145
        4     0      4       88   37.831784  44.168474  0.195851
        5     1      0       89    9.256317  37.884044  0.403487
        6     1      1       89   38.264777  30.120083  0.101970
        7     1      2       88   93.984462  36.118807  0.004633
        8     1      3       88  107.492325  51.553306  0.018531
        9     1      4       88  -16.508723  46.352634  0.639138
        10    2      0       89   31.201446  29.410963  0.144373
        11    2      1       89   27.746800  46.635800  0.275932
        12    2      2       88   87.038590  34.676231  0.006036
        13    2      3       88   -0.136508  55.222017  0.500986
        14    2      4       88   88.479385  32.922866  0.003600
        15    3      0       89  106.907846  53.418722  0.022679
        16    3      1       89   58.039351  37.127359  0.058997
        17    3      2       88   33.258400  31.773060  0.147608
        18    3      3       88   84.780801  35.499214  0.008465
        19    3      4       88   -6.289795  51.571162  0.548536

Breifly speaking, `cross_fit` controls aggregating results over `n_folds` given a repeat and `combine` controls combining results over `n_splits`. If both `cross_fit` and `combine` are `True`, the `summarize` method first aggregates the results of folds for each repeat and then do $p$-value combinations based on the results made by first step.

Finally, it is possible to combine all $p$-values without doing cross-fitting first by setting `combine = True` and `cross_fit = False`. This approach might result in conservative results for some p-value combination methods. However, so far it is hard to say which method is more reliable. The relative performances of these combination methods require further empirical evaluations.

In [17]:
cpi.summarize(
    cross_fit = False,
    combine = True)

Inferer Summary (cross_fit=False, combine=True)
 + Method: CPI (double_split=None, perturb_size=None)
 + Null Distribution: Normality (n_copies=100, n_permutations=None)
 + Loss Function: Mean Squared Error (reverse=False)


size   estimate  std_error   p_value
removal method                                      
sex     gmean   88.4  48.913911  39.509517  0.166355
        median  88.4  48.913911  39.509517  0.246342
        q1      88.4  48.913911  39.509517  0.035997
        min     88.4  48.913911  39.509517  0.022658
        hmean   88.4  48.913911  39.509517  0.088600
        hommel  88.4  48.913911  39.509517  0.081519
        cauchy  88.4  48.913911  39.509517  0.010957

## References

Tansey, W., Veitch, V., Zhang, H., Rabadan, R., & Blei, D. M. (2022). The holdout randomization test for feature selection in black box models. Journal of Computational and Graphical Statistics, 31(1), 151–162. doi: 10.1080/10618600.2021.1923520

Williamson, B. D., Gilbert, P. B., Simon, N. R., & Carone, M. (2023). A general framework for inference on algorithm-agnostic variable importance. Journal of the American Statistical Association, 118(543), 1645–1658. doi: 10.1080/01621459.2021.2003200